docs: adr auditable release (#2912)
kyma-project/telemetry-manager: docs/contributor/arch/031-auditable-release-automation.md

---
title: Auditable Release Automation
status: Proposed
date: 2026-01-20
---

# Auditable Release Automation

## Context

In the context of the SAP BTP, Kyma runtime product, ensuring a reliable and auditable release process is crucial for maintaining software quality and compliance. An automated release pipeline not only streamlines the deployment process but also provides a clear audit trail of changes, approvals, and deployments.

This document outlines the principles and practices for implementing auditable release automation within the Kyma runtime product environment.

## Principles of Auditable Release Automation

For auditing purposes, each release must provide successful test result artifacts. This includes:

- Test execution reports (unit and e2e tests)
- Docker images used for testing and release
- Gardener tests execution reports
- Image digest verification

### Test Execution Reports

> Test execution reports should be downloaded for both unit tests and end-to-end (E2E) tests from corresponding GitHub Workflow executions and retained for audit purposes.

Author: applied

### Deterministic Docker Builds
To achieve deterministic Docker builds and ensure audit compliance, we considered the following strategies:

- **Reproducible Build Tools**: Utilize tools and techniques that support reproducible builds, such as Docker's BuildKit with deterministic settings. This approach would require implementation support from the `Image-Builder` team and is not currently available.
- **Copy Release PR Image**: Modify the release process to copy the Docker image built during the PR directly to the production registry with the release tag. This approach ensures that the exact same image used for testing is released, maintaining identical digests.

**Recommended Approach:**

Due to the current unavailability of reproducible build tools, the recommended approach is, all tests (unit, E2E, Gardener, and upgrade tests) will be executed in the release branch using the release image before final release publication.

This approach provides:
- Solid image digests proof for testing and release (audit compliance)
- Comprehensive test coverage in the release context
- Clear traceability for release audits

![Release Workflow](./../assets/auditable-release-final.drawio.svg)

> ### The Module Release Publishing Workflow

Reviewer: ```suggestion
### Module Release Workflow
```

A separate GitHub workflow will be responsible for publishing the module release. This workflow will be triggered by the release master once the release version and release channel are entered. The workflow will handle the module config creation and channel assignment, and will be responsible for creating the release PRs for the `experimental`, `fast`, and `regular` channels.

![Module Release Workflow](./../assets/auditable-release-module-release.drawio.svg)

### The Management Plane Chart Release Workflow

A separate GitHub workflow will be responsible for bumping Management Plane Charts. This workflow will be triggered by the release master once the release version and the target chart are specified. The workflow will handle the chart release process for `chart/telemetry` and `chart/runtime-monitoring-operator` Helm charts.

![Management Plane Chart Release Workflow](./../assets/auditable-release-mpc-release.drawio.svg)

## Release Workflow Step-by-Step Execution

> 

Reviewer: Does this section try to provide details on the `![Release Workflow](./../assets/auditable-release-final.drawio.svg)` diagram? If yes, why aren't the diagram and the explanations next to each other?

**Project Master Action**: Close the current development milestone to signify the boundary between development and release phases. This marks the completion of all planned features for the current release.

**Release Master Action**: The release master initiates the release process by entering the release version and OpenTelemetry Collector Components (OCC) version for the release. This action triggers the release workflow and sets the stage for subsequent steps.

**System Decision Point**: The system evaluates whether this is a patch release or a new minor version release:
- **Patch Release**: If the release involves only bug fixes (patch release), skip release branch creation.
- **New Version Release**: If this is a new feature release (non-patch), create a dedicated release branch following the `release-x.y` naming convention (e.g., `release-1.0`).
- Commit these changes to the release branch
- Create a release tag marking the version point
**Source Control Action**: Push the committed version-bumped artifacts and the release tag to the release branch. This officially marks the release version in the repository.

> - Run Unit Tests and Release Image creation in parallel.

Reviewer: ```suggestion
6. The system builds the release Docker image and runs unit tests in parallel.
```

> **Simultaneous Test Runs**: Three independent test suites are triggered in parallel against the release branch after Docker image created:

Reviewer: ```suggestion
7. After the image is built, the system triggers the E2E, Gardener, and upgrade test suites in parallel. All tests run against the same release image to ensure consistency.
```

- **E2E (End-to-End) Tests**: Full system behavior verification
- **Gardener Integration Tests**: Compatibility validation with Gardener infrastructure platform
- **Upgrade Tests**: Validation of upgrade paths from previous versions to the new release

All tests execute against the same release Docker image to ensure reproducibility and consistency. Tests run in parallel to minimize total release cycle time.

**Completion Action**: Upon successful completion of all test suites, the system:
- Aggregates all test reports from the test workflows
- Downloads test execution logs and artifacts
- Uploads the complete test results to a pre-configured Google Cloud Storage (GCP) bucket

> **Release Publication**:

Reviewer: ```suggestion
9. The system creates an official GitHub release, attaching the relevant release artifacts.
```

## Conclusion

Implementing auditable release automation is essential for maintaining the integrity and compliance of the SAP BTP, Kyma runtime product. 
By ensuring that test reports are retained and that Docker images are reproducible, we can create a transparent and reliable release process that meets auditing requirements and enhances overall software quality.

The proposed strategies for deterministic Docker builds and the structured release workflow provide a clear path forward for achieving auditable release automation in the Kyma runtime product environment.

Currently, the recommended approach is to run all tests in the release branch using the release Docker image, ensuring that the same image is used for both testing and release, thus maintaining identical digests and providing a clear audit trail.

We can skip the PR tests and run them only in the release branch, so we can automate the release process without waiting for the PR tests to complete, and still ensure that the released image is tested and has the same digest as the one built in the PR.

> A new GitHub Action can be implemented to trigger the release branch workflow once the release master enters the release version and OpenTelemetry Collector Components version for the release.

Reviewer: Why "can be implemented"? Isn't this a decision record, meaning that the decision was made to implement the new workflow?

```suggestion
We implement a new GitHub Action that triggers the release branch workflow when the release master enters the release version and OpenTelemetry Collector Components version for the release.
```

## Implementing Auditable Release Automation

### Download Test Reports

> The test reports from unit tests, E2E tests, and Gardener tests should be downloaded from the respective GitHub Actions workflows and stored as artifacts for audit purposes. For this pupose, a new re-usable workflow created to download and store test reports based on workflow run ID and job name.

Author: updated

The new workflow can be called from the release PR workflow after test jobs are completed successfully and upload to the pre-configured GCP bucket for audit retention as desccribed [here](https://github.tools.sap/kyma/backlog/issues/8419).

### Deterministic Docker Builds
To achieve deterministic Docker builds, the following strategies can be employed:

> - **Reproducible Build Tools**: Utilize tools and techniques that support reproducible builds, such as Docker's BuildKit option, which have to be implemented by the `Image-Builder`.

Reviewer: ```suggestion
- **Reproducible Build Tools**: Utilize tools and techniques that support reproducible builds, such as Docker's BuildKit option, which must be implemented by the `Image-Builder`.
```

![Release Workflow](./../assets/auditable-release-final.drawio.svg)

### Module Release Workflow

> After entering the release version and channel, the release master triggers a dedicated GitHub workflow to publish the module release. The workflow then creates the module configuration, assigns the release channel, and opens pull requests to update the \experimental`, `fast`, and `regular` channels.`

Reviewer: Just to follow up on [open questions](https://github.com/kyma-project/telemetry-manager/pull/2912#discussion_r2816315888):
**Is it "After entering..." or "By entering..." ?**

Here, you chose "after", but in line 87 you use "by". 
This distinction matters, because "by" means that entering the data triggers a workflow, while "after" means that the user has to trigger the workflow manually.

![Module Release Workflow](./../assets/auditable-release-module-release.drawio.svg)

### Management Plane Chart Update Workflow

After specifying the release version and a target chart, the release master triggers a dedicated GitHub workflow to bump Management Plane Charts. This workflow automates the release of charts such as \chart/telemetry` and `chart/runtime-monitoring-operator`.`

![Management Plane Chart Release Workflow](./../assets/auditable-release-mpc-release.drawio.svg)

## Release Workflow Steps

1. The Project Master closes the current development milestone to mark the end of the development phase for the release.

2. The Release Master triggers the release workflow by entering the release version and the OpenTelemetry Collector Components (OCC) version.

3. The system evaluates if the release is a patch or a new minor version: 
   - For a new minor version, it creates a dedicated `release-x.y` branch.
   - For a patch, it skips release branch creation.
4. The system commits the version bump and creates a release tag.

> **Source Control Action**: Push the committed version-bumped artifacts and the release tag to the release branch. This officially marks the release version in the repository.

Reviewer: We seem to have a mix of the old version and the numbered steps here. Please review so that it's consistent:

<img width="583" height="710" alt="Image" src="https://github.com/user-attachments/assets/0dfbd21e-1758-43df-9d88-992b11e7deec" />
